extensions/amp-ad-exit/0.1/amp-ad-exit.js

```diff
@@ -100,17 +103,68 @@ export class AmpAdExit extends AMP.BaseElement {
       'CLICK_Y': true,
     };
     if (target.vars) {
-      for (const customVar in target.vars) {
-        if (customVar[0] == '_') {
-          vars[customVar] = () =>
-              args[customVar] || target.vars[customVar].defaultValue;
-          whitelist[customVar] = true;
+      const all3pResponses = this.getAmpDoc().win[AMP_ANALYTICS_3P_RESPONSES] ||
+          {};
+
+      for (const customVarName in target.vars) {
+        if (customVarName[0] == '_') {
+          const customVar =
+              /** @type {./config.Variable} */ (target.vars[customVarName]);
+          if (customVar) {
+            /*
+              Example:
+              The amp-ad-exit target has a variable representing the
+               priority of something, which is defined as follows:
+               "vars": {
+                 "_pty": {
+                   "defaultValue": "unknown",
+                   "vendorAnalyticsSource": "vendorXYZ",
+                   "vendorAnalyticsResponseKey": "priority"
+                 },
+                 ...
+               }
+               The cross-domain iframe of vendorXYZ has sent the
+               following response for the creative:
+                 { priority: medium, category: W }
+               This is just example data. The keys/values in that object can
+               be any strings.
+               The code below will create substitutionFunctions['_pty'],
+               which in this example will return "medium".
+             */
+            substitutionFunctions[customVarName] = () => {
+              if (customVar.hasOwnProperty('vendorAnalyticsSource') &&
+                  customVar.hasOwnProperty('vendorAnalyticsResponseKey')) {
+                // It's a 3p analytics variable
+                const vendor =
+                    /** @type {string} */ (customVar.vendorAnalyticsSource);
+                if (all3pResponses[vendor]) {
+                  /* The vendor (in the example above, "vendorXYZ") has
+                     responded to this creative. Need to check whether that
+                     response contains a property that matches the
+                     vendorAnalyticsResponseKey (ex: "priority") for this
+                     custom variable. If so, return the value in the
+                     response object that is associated with that key.
+                  */
+                  const relevant3pResponses = all3pResponses[vendor];
+                  if (relevant3pResponses) {
+                    return relevant3pResponses[
+                        /** @type {string} */
+                        (customVar.vendorAnalyticsResponseKey)];
+                  }
+                }
+              }
+              // Either it's not a 3p analytics variable, or it is one but
+              // no matching response has been received yet.
```

Comment:
What should happen if the vendor JS sends responses with different keys (if that's allowed)? Suppose that the first message is {fastVerification: 'abc'} and later the iframe sends {fastVerification: 'abc', slowVerification: '123'}, and the exit variable only wants "slowVerification". Then should this variable replacement insert undefined, as it does now (with the return on L150), or fall back to this section?

Reply:
So, let's say the config contains:
```
...
finalUrl:"http://somewhere/?slow=_sv&fast=_fv",
vars: {
  _sv: {
    vendorAnalyticsSource: "foo",
    vendorAnalyticsResponseKey: "slowVerification",
    defaultValue: "000"
  },
  _fv: {
    vendorAnalyticsSource: "foo",
    vendorAnalyticsResponseKey: "fastVerification",
    defaultValue: "999"
  }
}
...
```
I would think that the order would/should go:
1) Receive {fastVerification: 'abc'}
2) If the user were to click now, they'd go to http://somewhere/?slow=000&fast=abc
3) Receive {fastVerification: 'abc', slowVerification: '123'}
4) If the user were to click now, they'd go to http://somewhere/?slow=123&fast=abc
(Note that the return on L150 applies to the function begun on L134, and does not break us out of the for loop on line 109)

---
On the other hand, if the messages were sent in the opposite order to what you just wrote (fast+slow, and then just fast), I would think that at this point slow should be gone, and trying to substitute it should give the defaultValue, so they'd go to http://somewhere/?slow=000&fast=abc in that case.

Reply:
Yep, you're right. Thanks for catching this error. Fixed.